### DISCUSSION TITLE
Functionality Question About Agents

### DISCUSSION THREAD

**Discussion by bc24fl** [narrowlink/narrowlink#93] [General]
My apologies if this was addressed in the docs and I've misunderstood it.  I'm a little confused about the agents and when it's needed.  It's my understanding that an Agent is used to proxy requests from clients (behind NAT).  But what if a client has a public IP address like a server on the cloud?  Will the client install suffice or would you also have to install the Agent software as well?  To add a little context, I've used other tech like Tailscale and an Agent (or Subnet Router as they call it) is only really needed when an installation is not possible (IOT, Printers, etc..).  Does NL work the same or is an Agent always required?

**Comment by SajjadPourali** [comment]
Narrowlink operates differently, functioning as a proxy rather than a VPN. In a VPN, both endpoints (like Computer A and B) can initiate and receive connections bidirectionally. However, in a proxy approach (like NL), the client can only instruct the agent to establish a connection, and the agent sends it on behalf of the client. Importantly, the client in NL client never receives new connections from the agent; it only receives responses to its requests.

It seems you are comparing NL to a VPN approach where a single client can handle bidirectional connectivity. To achieve such a feature in NL, you need both the client and agent to be running on the machine for bidirectional communication due to its unidirectional nature, regardless of whether your device has a valid IP address or not.

In a nutshell, based on your context, Narrowlink has one client and one subnet router (agent) that allows you to connect to local or remote machines. To connect your client to the agent machine, not its subnet network, you can use the 127.0.0.1 address.
Utilizing a client without an agent is like having a remote control without a device to operate.

**Tun Mode Example:**
```bash
sudo narrowlink t -gn agent_name -m 10.10.10.10=127.0.0.1 # redirects 10.10.10.10 to the machine that installed agent
```
**Port Forwarding Example**
```bash
narrowlink f -n agent_name -l 127.0.0.1:8080 127.0.0.1:80 # opens port 8080 on the client machine and forward them to the agent machine at address 127.0.0.1:80 which is port 80 of the agent
```

Please don't hesitate to tell me if any further elaboration is required.

-------

P.S. There is a workaround for Narrowlink 0.3 to make it more user-friendly.
